clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
+        return STATUS_ERROR_INVALID_PARAMETER;
+
+    schedule_entry_t entry(0, 0, 0, 0);
+    // See comment in read_recorded_schedule() on our assumption that we can
+    // easily fit the whole context switch sequence in memory.
+    // We do not have a subfile listing feature in archive_istream_t, but we can
+    // read sequentially as each record has a cpu field.
+    // This schedule_entry_t format doesn't have the stop instruction ordinal (as it was
+    // designed for skip targets only), so we take two passes to get that information.
+    // We also need to translate the thread and cpu id values into 0-based ordinals.
+    std::unordered_map<memref_tid_t, input_ordinal_t> tid2input;
+    for (int i = 0; i < static_cast<input_ordinal_t>(inputs_.size()); ++i) {
+        tid2input[inputs_[i].tid] = i;
+    }
+    std::unordered_map<uint64_t, output_ordinal_t> cpu2output;
+    std::vector<std::map<uint64_t, uint64_t>> start2stop(inputs_.size());
+    // We number the outputs according to their order in the file.
+    // XXX: Should we support some direction from the user on this?
+    // What about dividing by socket: we may want to do that using the cpuid.
+    output_ordinal_t cur_output = 0;
+    uint64_t cur_cpu = std::numeric_limits<uint64_t>::max();
+    // We also want to collapse same-cpu consecutive records so we start with
+    // a temporary local vector.
+    std::vector<std::vector<schedule_record_t>> all_sched(outputs_.size());
+    while (options_.replay_as_traced_istream->read(reinterpret_cast<char *>(&entry),
+                                                   sizeof(entry))) {
+        if (entry.cpu != cur_cpu) {
+            if (cur_cpu != std::numeric_limits<uint64_t>::max()) {
+                ++cur_output;
+                if (cur_output >= static_cast<int>(outputs_.size())) {
+                    error_string_ = "replay_as_traced_istream cpu count != output count";
+                    return STATUS_ERROR_INVALID_PARAMETER;
+                }
+            }
+            cur_cpu = entry.cpu;
+            cpu2output[entry.cpu] = cur_output;
+        }
+        input_ordinal_t input = tid2input[entry.thread];
+        // We'll fill in the stop ordinal in our second pass below.
+        uint64_t start = entry.instr_count;
+        uint64_t timestamp = entry.timestamp;
+        // Some entries have no instructions.
+        if (all_sched[cur_output].size() > 0 &&
+            input == all_sched[cur_output].back().key.input &&
+            start == all_sched[cur_output].back().start_instruction) {
+            VPRINT(this, 3,
+                   "Output #%d: as-read segment #%zu has no instructions: skipping\n",
+                   cur_output, all_sched[cur_output].size() - 1);
+            continue;
+        }
+        all_sched[cur_output].emplace_back(schedule_record_t::DEFAULT, input, start, 0,
+                                           timestamp);
+        start2stop[input][entry.instr_count] = entry.instr_count;
```

Comment:
Perhaps a `set` is what we need?

Reply:
Yes, we just need a sorted container where we can ++ to find the next element.  Changed.